modules/magnolia/src/main/scala/pureconfig/module/magnolia/MagnoliaConfigReader.scala

```diff
@@ -0,0 +1,94 @@
+package pureconfig.module.magnolia
+
+import _root_.magnolia._
+import pureconfig.ConfigReader.Result
+import pureconfig._
+import pureconfig.error.{ ConfigReaderFailures, KeyNotFound, UnknownKey, WrongSizeList }
+import pureconfig.generic.{ CoproductHint, ProductHint }
+
+/**
+ * An object containing Magnolia `combine` and `dispatch` methods to generate `ConfigReader` instances.
+ */
+object MagnoliaConfigReader {
+
+  def combine[A](ctx: CaseClass[ConfigReader, A])(implicit hint: ProductHint[A]): ConfigReader[A] =
+    if (ctx.typeName.full.startsWith("scala.Tuple")) combineTuple(ctx)
+    else if (ctx.isValueClass) combineValueClass(ctx)
+    else combineCaseClass(ctx)
+
+  def combineCaseClass[A](ctx: CaseClass[ConfigReader, A])(implicit hint: ProductHint[A]): ConfigReader[A] = new ConfigReader[A] {
+    def from(cur: ConfigCursor): Result[A] = {
+      cur.asObjectCursor.flatMap { objCur =>
+        val res = ctx.constructMonadic[Result, Param[ConfigReader, A]#PType] { param =>
+          val keyStr = hint.configKey(param.label)
+          objCur.atKeyOrUndefined(keyStr) match {
+            case keyCur if keyCur.isUndefined =>
+              param.default match {
+                case Some(defaultValue) if hint.useDefaultArgs => Right(defaultValue): Result[param.PType]
+                case _ if param.typeclass.isInstanceOf[ReadsMissingKeys] => param.typeclass.from(keyCur)
+                case _ => cur.failed(KeyNotFound.forKeys(keyStr, objCur.keys))
+              }
+            case keyCur =>
+              param.typeclass.from(keyCur)
+          }
+        }
+        if (hint.allowUnknownKeys || res.isLeft) res
+        else {
+          val usedKeys = ctx.parameters.map { param => hint.configKey(param.label) }.toSet
+          val unknownKeyFailures = objCur.map.collect {
+            case (k, keyCur) if !usedKeys(k) => keyCur.failureFor(UnknownKey(k))
+          }.toList
+
+          if (unknownKeyFailures.isEmpty) res
+          else Left(ConfigReaderFailures(unknownKeyFailures.head, unknownKeyFailures.tail))
+        }
+      }
+    }
+  }
+
+  def combineTuple[A](ctx: CaseClass[ConfigReader, A]): ConfigReader[A] = new ConfigReader[A] {
+    def from(cur: ConfigCursor): Result[A] = {
+      val collCur = cur.asListCursor.map(Right.apply).left.flatMap(failure =>
+        cur.asObjectCursor.map(Left.apply).left.map(_ => failure))
+
+      collCur.flatMap {
+        case Left(_) => combineCaseClass(ctx).from(cur)
```

Comment:
I suppose we can also use the cursor on the `Left` instead of `cur`. It'd avoid an extra unnecessary conversion to a `ConfigObjectCursor`.

Reply:
> This `combineCaseClass` is relying on the default `ProductHint` implementation. While it's very unlikely that we'll ever change it in a way that would break this, I think it'd be safer to provide an explicit `ProductHint` that we know won't mess with the field names.

Good point, I didn't think of that. The current generic derivation behavior is to fall back to reading the tuple exactly like you would read a case class (including hints), so the matching behavior here is actually to also propagate `ProductHint` from `combine` to `combineTuple`.

> Also, I think it's preferable to cache the call to combineCaseClass(ctx) outside the from method, to avoid creating a new ConfigReader on every from call.

This is a micro optimization where we're sacrificing extra persistent memory for a minimal increase in performance. We're trading a simple object instantiation (the only thing `combineCaseClass` does) to have a type class with an extra field in its object, something I don't usually consider worth it.

> I suppose we can also use the cursor on the Left instead of cur. It'd avoid an extra unnecessary conversion to a ConfigObjectCursor.

Sounds good :)